cmd/olympus/actions/merge_db.go

```diff
@@ -0,0 +1,96 @@
+package actions
+
+import (
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/gomods/athens/pkg/storage"
+
+	"github.com/gomods/athens/pkg/cdn"
+	"github.com/gomods/athens/pkg/eventlog"
+)
+
+// mergeDB merges diff into the module database.
+//
+// TODO: this is racey if multiple processes are running mergeDB (they will be!) in a few ways:
+//
+// 1. CDN updates that race to change the /list endpoint
+// 2. races between CDN updates and module metadata updates. For example:
+//		- Delete operation deletes from the CDN
+//		- Add operation adds to the CDN and saves to the module metadata DB
+//		- Delete operation adds tombstone to module metadata k/v store
+//
+// Both could be fixed by putting each 'for' loop into a (global) critical section
+func mergeDB(originURL string, diff dbDiff, eLog eventlog.Eventlog, storage storage.Backend) error {
+	for _, added := range diff.Added {
+		if _, err := eLog.ReadSingle(added.Module, added.Version); err != nil {
+			// the module/version already exists, is deprecated, or is
+			// tombstoned, so nothing to do
+			continue
+		}
+
+		// download code from the origin
+		data, err := cdn.Download(originURL, added.Module, added.Version)
+		if err != nil {
+			log.Printf("error downloading new module %s/%s from %s (%s)", added.Module, added.Version, originURL, err)
+			continue
+		}
+
+		// save module data to the CDN
+		zipBytes, err := ioutil.ReadAll(data.Zip)
+		if err != nil {
+			continue
+		}
+		if err := data.Zip.Close(); err != nil {
+			continue
+		}
+
+		if err := storage.Save(added.Module, added.Version, data.Mod, zipBytes, data.Info); err != nil {
```

Comment:
Does the CDN Storage implement the storage.Backend interface? If that's the case do we need the cdn.Saver interface?

Reply:
few changes we discussed with aaron here:
- there is no CDN storage, just regular storage and CDN attached to it. just like azure cdn to azure blob.
- we will have CDN Metadata K/V which holds redirect uris
- we will have eventlog which holds module,version,operation triplet, eventlog is append only

Reply:
i will need to clean unneeded code as well. probably in separate PR

Reply:
One more question - so I guess the storage.backend you're passing here is the aws S3/azure blob storage and so on, right?
Where will the CDN metadata (redirect URL) be created?

Reply:
i guess on package processing based on configuration.
imagine you process a module, downloads sources uploads to s3 as `/mod/@v/v1.0` so CDN link stored will be env["CDN_DOMAIN"] + "/mod/@v/v1.0" = `aws-cdn/mod/@v/v1.0`
at this point you dont even need to store that as it can be computed.

now imagine your you will start using different storage, and keep old one as is. new packages goes to azure blob storage as `/mod/@v/v2.0` configuration is changed so CDN domain points to azure CDN and URI will be updated to env["CDN_DOMAIN"] + "/mod/@v/v2.0" = `azure-cdn/mod/@v/v1.0`

so now metadata for
- v1 says `aws-cdn/mod/@v/v1.0`
- v2  says `azure-cdn/mod/@v/v1.0`
both works.

you can easily delete aws if it is deprecated which will lead to cache miss and repopulating with azure backing storage.

does this make sense?

Reply:
In https://github.com/gomods/athens/pull/170 I built the CDN driver for Azure and ended up removing the `cdn.Saver` interface ([here](https://github.com/gomods/athens/pull/170/files#diff-7b85c056fb0e17fbf182ac3616e0f2ca)). 

`pkg/cdn` can use `pkg/storage`'s saver. the other stuff (getter, lister, etc...) is only used by the proxy, because that's the only thing that actually serves source code (and `go.mod` and stuff). I implemented the Azure CDN getter (the thing that returns the redirect URL) and Azure storage Saver separately, and then composed the storage Saver into the CDN getter [here](https://github.com/gomods/athens/blob/353e1d1f619d611da20fa69a6735712b6713b678/pkg/cdn/metadata/azurecdn/getter.go)